plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,849 @@
+	if (fw_sz == acksz)
+		return TRUE;
+
+        g_prefix_error(error,
+                       "device nack the incoming filesize, requested: %lu, ack: %lu",
+                       fw_sz,
+                       acksz);
+	return FALSE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_set_data(FuMediatekScalerDevice *self, FuChunk *chk, GError **error)
+{
+	g_autoptr(GPtrArray) chk_slices = NULL;
+	g_autoptr(GBytes) chk_bytes = fu_chunk_get_bytes(chk);
+	const guint8 buf[2] = {DDC_OPCODE_MTK_REQ, DDC_VCP_CODE_SET_DATA};
+
+	chk_slices = fu_chunk_array_new_from_bytes(chk_bytes, 0x00, 0x00, DDC_DATA_FRAGEMENT_SIZE);
+	for (guint i = 0; i < chk_slices->len; i++) {
+		g_autoptr(GByteArray) cod_setdata = g_byte_array_new();
+		FuChunk *chk_slice = g_ptr_array_index(chk_slices, i);
+
+		g_byte_array_append(cod_setdata, buf, sizeof(buf));
+		g_byte_array_append(cod_setdata,
+				    fu_chunk_get_data(chk_slice),
+				    (guint)fu_chunk_get_data_sz(chk_slice));
+		if (!mediatek_scaler_device_ddc_write(self, cod_setdata, error)) {
+			g_prefix_error(error, "failed to send firmware to device: ");
+			return FALSE;
+		}
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_get_staged_data(FuMediatekScalerDevice *self, guint16 *chksum, guint32 *pktcnt, GError **error)
+{
+	g_autoptr(GByteArray) cmdbuf = g_byte_array_new();
+	g_autoptr(GByteArray) readbuf = g_byte_array_new();
+	const guint8 cmd[] = {DDC_OPCODE_MTK_REQ, DDC_VCP_CODE_GET_STAGED};
+
+	g_byte_array_append(cmdbuf, cmd, sizeof(cmd));
+	if (mediatek_scaler_device_ddc_read(self, cmdbuf, readbuf, error))
+		return FALSE;
+
+	if (readbuf->len >= 7) {
+                *chksum = fu_memread_uint16(&readbuf->data[2], G_LITTLE_ENDIAN); /* 2 bytes */
+                *pktcnt = fu_memread_uint32(&readbuf->data[4], G_LITTLE_ENDIAN); /* 4 bytes */
+                return TRUE;
+        }
+        g_set_error(error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "invalid staged data, length of read: %u", readbuf->len);
+        return FALSE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_check_sent_info(FuMediatekScalerDevice *self,
+					  GBytes *fwdata,
+					  GError **error)
+{
+	for (gint retry = 0; retry < DDC_RW_MAX_RETRY_CNT; retry++) {
+                guint16 chksum = 0;
+                guint32 pktcnt = 0;
+
+                /* delay before the read attempt */
+		fu_device_sleep(FU_DEVICE(self), 10);
+
+                if(!fu_mediatek_scaler_device_get_staged_data(self, &chksum, &pktcnt, error))
+                        return FALSE;
+
+                /* verify the packets on chip and checksum */
+		if (g_bytes_get_size(fwdata) != pktcnt) {
+			g_debug("failed data packets verification, sent size: %lu, ack size: %u",
+				g_bytes_get_size(fwdata),
+				pktcnt);
+		} else if (fu_sum16_bytes(fwdata) != chksum) {
+			g_debug("failed data checksum comparison, expected: %u, got: %u",
+				fu_sum16_bytes(fwdata),
+				chksum);
+		} else
```

Comment:
useless else

Reply:
it returns `true` if it passes the two checks, otherwise continuously retry the data until it reached the maximum